File: go/0700E- 二叉搜索树中的搜索.go
/*
 * @Author: SukiEva
 * @Date: 2021-11-10 19:47:59
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func searchBST(root *TreeNode, val int) *TreeNode {
	for root != nil {
		if root.Val > val {
			root = root.Left
		} else if root.Val < val {
			root = root.Right
		} else {
			return root
		}
	}
	return nil
}

func searchBSTOld(root *TreeNode, val int) *TreeNode {
	if root == nil || root.Val == val {
		return root
	}
	if root.Val > val {
		return searchBSTOld(root.Left, val)
	}
	if root.Val < val {
		return searchBSTOld(root.Right, val)
	}
	return nil
}

File: cpp/0700E-二叉搜索树中的搜索.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-22 10:16:41
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    TreeNode* searchBST(TreeNode* root, int val) {
        if (root == nullptr) return nullptr;
        if (root->val < val) return searchBST(root->right, val);
        if (root->val > val) return searchBST(root->left, val);
        return root;
    }
};